crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,397 @@
+        );
+
+        let latest = LatestSnapshot {
+            version: snapshot.version,
+            cluster_name: snapshot.cluster_name.clone(),
+            node_name: snapshot.node_name.clone(),
+            partition_id: snapshot.partition_id,
+            snapshot_id: snapshot.snapshot_id,
+            created_at: snapshot.created_at.clone(),
+            min_applied_lsn: snapshot.min_applied_lsn,
+            path: relative_snapshot_path,
+        };
+        let latest_path = object_store::path::Path::from(format!(
+            "{prefix}{partition_id}/latest.json",
+            prefix = self.prefix,
+            partition_id = snapshot.partition_id,
+        ));
+
+        // By performing a CAS on the latest snapshot pointer, we can ensure strictly monotonic updates.
+        let maybe_stored = match self.object_store.get(&latest_path).await {
+            Ok(result) => {
+                let attributes = result.attributes.clone();
+                let parse_result: serde_json::Result<LatestSnapshot> =
+                    serde_json::from_slice(result.bytes().await?.iter().as_slice());
+                parse_result
+                    .inspect_err(|e| {
+                        info!(
+                            repository_latest_lsn = "unknown",
+                            new_snapshot_lsn = ?snapshot.min_applied_lsn,
+                            "Failed to parse stored latest snapshot pointer, will update it: {}",
+                            e
+                        )
+                    })
+                    .ok()
+                    .map(|metadata| (attributes, metadata))
+            }
+            Err(object_store::Error::NotFound { .. }) => {
+                debug!(
+                    repository_latest_lsn = "none",
+                    new_snapshot_lsn = ?snapshot.min_applied_lsn,
+                    "No latest snapshot pointer found, will create one"
+                );
+                None
+            }
+            Err(e) => {
+                bail!("Failed to get latest snapshot pointer: {}", e);
+            }
+        };
+
+        if maybe_stored
+            .as_ref()
+            .is_some_and(|(_, stored)| stored.min_applied_lsn >= snapshot.min_applied_lsn)
+        {
+            let repository_latest_lsn = maybe_stored.expect("is some").1.min_applied_lsn;
+            info!(
+                ?repository_latest_lsn,
+                new_snapshot_lsn = ?snapshot.min_applied_lsn,
+                "Newly created snapshot is not newer than the latest stored snapshot, will not update latest pointer"
+            );
+            return Err(anyhow!(
+                "Snapshot repository already contains snapshot at LSN {}",
+                repository_latest_lsn,
+            ));
+        }
+
+        let latest_json = PutPayload::from(serde_json::to_string_pretty(&latest)?);
+        let conditions = maybe_stored
+            .map(|(attributes, _)| PutOptions {
+                mode: PutMode::Overwrite,
```

Comment:
The docs of `Overwrite` say that this operation will overwrite any value stored on the same part. Don't we want to have a conditional put that checks the etag version? In this case, I think you need to use `PutMode::Update(etag)`.

Reply:
Yes, we do! I completely messed up the API usage here - updated revision coming soon.